src/abstractdataframe/iteration.jl

```diff
@@ -391,7 +406,21 @@ function mapcols(f::Union{Function, Type}, df::AbstractDataFrame)
             push!(vs, [fv])
         end
     end
-    return DataFrame(vs, _names(df), copycols=false)
+
+    new_df = DataFrame(vs, _names(df), copycols=false)
+
+    _copy_metadata!(new_df, df)
+
+    # in DataFrames.jl parent always returns source DataFrame
+    if getfield(parent(df), :colmetadata) !== nothing && (f === identity || f === copy)
```

Comment:
we discussed it in DataAPI.jl as far as I remember and decided not to add it.
I think it is OK to add it later if needed. Normal users probably will not need it.

Reply:
I could omit it as it is only needed for performance.

Reply:
My point is that if we need it here, other packages are likely to want to use a similar optimization at some point. We could add it to DataAPI later, but then it's tricky to know when you can rely on it as depending on a recent DataAPI version isn't enough to ensure that packages implementing the metadata interface also implement that function, as we learnt at https://github.com/JuliaData/DataAPI.jl/pull/46.

Reply:
OK - let us discuss it. Hadding `hascolmetadata(df)` returning `true` if any column has any metadata is not a problem.
The problem is that then it would be natural to add `colmetadata(df)` and it is not clear what data structure should be returned by this call. But maybe it is enough to define `hascolmetadata(df)` in DataAPI.jl without defining `colmetadata(df)`. What do you think?